### DISCUSSION TITLE
Data Importer - Nordigen API is configured incorrectly

### DISCUSSION THREAD

**Discussion by chazman88** [firefly-iii/firefly-iii#7237] [Firefly III]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [ ] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [ ] ... there are no open or closed issues that are related to my problem
- [ ] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

Nordigen importer failed with a Server 500 Message. 

I now have "The Nordigen API is configured incorrectly and cannot be used to import data." where the Import From Nordigen button should be.

Running both Firefly and importer as Dockers on UNRAID. Have tried rebooting the importer docker but still have same error. 

### Debug information

Debug information generated at 2023-03-15 11:01:32 Europe/Madrid for Firefly III Data Importer version **1.2.1**.

| Scope | Version |
| --- | --- |
| FIDI | 1.2.1 |
| Build | 201, 12-03-2023 18:16:05 CET |
| PHP | 8.2.3 |
| Host | Linux |

| System info | Value |
| --- | --- |
| Using docker? | true |
| System TZ | Europe/Madrid |
| Browser TZ | Europe/London |
| App environment | production |
| App debug mode | false |
| App cache driver | file |
| App logging | emergency, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Interface | apache2handler |
| BCscale | 12 |
| Trusted proxies (.env) |  |

| User info | Value |
| --- | --- |
| User agent | Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.3 Safari/605.1.15 |

**Comment by JC5** [comment]
Well if the data importer is configured incorrectly, this isn't really a surprise 😉 . What are you trying to accomplish?

**Reply by chazman88** [reply]
Was trying to import data from my bank. 

My normal process was to boot up the Importer Docker, hit the import from nordigen button and then fill in all the details each time for an import. 
After something went wrong and I got the Server Error 500 then the button disappeared and was replaced with the error. I had no idea how to clear the error so that I could start the process again. 

However, just booting back into the importer now the error has cleared and I have the Import from Nordigen button back. So looks like it was a time based thing before the error cleared itself?

**Reply by JC5** [reply]
That could be it. Or Nordigen had a hick-up perhaps?

**Reply by Faust404** [reply]
I'm having the exact same issue right now. It's been almost 24 hours and the issue still persists. Is there a solution to this? I have tried restarting the container several times as well. I also provided the nordigen id and secret in the env file. That did not help either. Are there any other things I could do to help with this?

**Reply by JC5** [reply]
There may be more details in your logs?

**Reply by Faust404** [reply]
I am using docker so are you referring to the logs of the firefly-importer container?

**Reply by JC5** [reply]
Yes, the logs of the importer.

**Reply by Faust404** [reply]
I have put the last 1000 lines of logs into pastebin. Please have a look
https://pastebin.com/1N3u6rnx

**Reply by JC5** [reply]
I see the same error again and again: "No active account found with the given credentials"

That's a response from Nordigen I can't help I'm afraid.

**Reply by Faust404** [reply]
That could be the case because I got a mail from Nordigen yesterday saying this:

> I hope you are enjoying testing Nordigen. 

> While testing the end user flow, you may have noticed the “not verified” banner on the consent step. An account not being verified doesn’t limit the API functionality in any way. Its only purpose is to urge end users to be cautious when connecting accounts to unknown entities meaning that if you intend to use the API for personal or internal use only, verification is not needed. 

> However, if you’re building a commercial application, your account should be verified so your users don't get discouraged from connecting their accounts.  

> If this is the case, please come back to me with information on your use case and I’ll help you complete the verification process. This process normally takes 1-2 business days - after that, you'll be ready to launch. 

Could this be related to what you are saying?

Btw I really appreciate that you are taking time to answer these queries. Thank you so much.

**Reply by rchl** [reply]
How to reset the configuration (where is it stored)?

I'm getting the same message and I think that's because I've generated a new requisition manually through the API after using it in the Importer once. So now the requisition used in the Importer is expired and I don't see a way to reset stored configuration and re-authenticate with Nordigen again.

**Reply by rchl** [reply]
I figured out that it's stored in `nordigen*` cookies so deleting those allowed me to re-authenticate.

**Reply by JC5** [reply]
Yes, that's also possible using the button [Reauthenticate]

**Reply by rchl** [reply]
No sorry, that didn't work. Tried both that button and "start over" and it kept showing same error.

**Reply by rchl** [reply]
And also, I've just opened Importer and it's showing the "The Nordigen API is configured incorrectly and cannot be used to import data." error again.

I've used Nordigen successfully yesterday once, after deleting cookies, but it seems like access expired now or something.

I know how to fix now so no big deal.

**Comment by nick2k3** [comment]
Hi, I am having a similar issue that is driving me crazy.
My internet provider recently changed my IP address and the ID/token pairs I had generated were strictly associated with my IP address with a /32 subnet.
I have regenerated a new pair of credential with the new IP/32 and updated the credential in the FIDI container.
So far no changes.

Here is what the logs say:
```
idation")  
[2023-06-15 20:11:12] production.DEBUG: Now at App\Services\Nordigen\AuthenticationValidator::validate  
[2023-06-15 20:11:12] production.DEBUG: Now at App\Services\Nordigen\TokenManager::hasValidRefreshToken  
[2023-06-15 20:11:12] production.DEBUG: No Nordigen refresh token, so return false.  
[2023-06-15 20:11:12] production.DEBUG: No refresh token, so return false.  
[2023-06-15 20:11:12] production.DEBUG: Now at App\Services\Nordigen\TokenManager::getNewTokenSet  
192.168.10.50 - - [15/Jun/2023:18:11:12 +0000] "GET /validate/spectre HTTP/1.1" 200 1147 "http://192.168.10.100:8098/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.5 Safari/605.1.15"
[2023-06-15 20:11:12] production.DEBUG: Result is {"result":"OK","message":null} 
192.168.10.50 - - [15/Jun/2023:18:11:12 +0000] "GET /token/validate HTTP/1.1" 200 1158 "http://192.168.10.100:8098/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.5 Safari/605.1.15"
[2023-06-15 20:11:12] production.ERROR: Client error: `POST https://ob.nordigen.com/api/v2/token/new/` resulted in a `401 Unauthorized` response:
summary=Authentication failed detail=No active account found with the given credentials status_code=401
 {"exception":"[object] (GuzzleHttp\\Exception\\ClientException(code: 401): Client error: `POST https://ob.nordigen.com/api/v2/token/new/` resulted in a `401 Unauthorized` response:
{\"summary\":\"Authentication failed\",\"detail\":\"No active account found with the given credentials\",\"status_code\":401}
 at /var/www/html/vendor/guzzlehttp/guzzle/src/Exception/RequestException.php:113)

[stacktrace]
#0 /var/www/html/vendor/guzzlehttp/guzzle/src/Middleware.php(72): GuzzleHttp\\Exception\\RequestException::create()
#1 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(209): GuzzleHttp\\Middleware::GuzzleHttp\\{closure}()
#2 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(158): GuzzleHttp\\Promise\\Promise::callHandler()
#3 /var/www/html/vendor/guzzlehttp/promises/src/TaskQueue.php(52): GuzzleHttp\\Promise\\Promise::GuzzleHttp\\Promise\\{closure}()
#4 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(251): GuzzleHttp\\Promise\\TaskQueue->run()
#5 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(227): GuzzleHttp\\Promise\\Promise->invokeWaitFn()
#6 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(272): GuzzleHttp\\Promise\\Promise->waitIfPending()
#7 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(229): GuzzleHttp\\Promise\\Promise->invokeWaitList()
#8 /var/www/html/vendor/guzzlehttp/promises/src/Promise.php(69): GuzzleHttp\\Promise\\Promise->waitIfPending()
#9 /var/www/html/vendor/guzzlehttp/guzzle/src/Client.php(189): GuzzleHttp\\Promise\\Promise->wait()
#10 /var/www/html/vendor/guzzlehttp/guzzle/src/ClientTrait.php(95): GuzzleHttp\\Client->request()
#11 /var/www/html/app/Services/Nordigen/Request/PostNewTokenRequest.php(62): GuzzleHttp\\Client->post()
#12 /var/www/html/app/Services/Nordigen/TokenManager.php(77): App\\Services\\Nordigen\\Request\\PostNewTokenRequest->post()
#13 /var/www/html/app/Services/Nordigen/AuthenticationValidator.php(66): App\\Services\\Nordigen\\TokenManager::getNewTokenSet()
#14 /var/www/html/app/Http/Controllers/ServiceController.php(55): App\\Services\\Nordigen\\AuthenticationValidator->validate()
#15 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Controller.php(54): App\\Http\\Controllers\\ServiceController->validateNordigen()
#16 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(43): Illuminate\\Routing\\Controller->callAction()
#17 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Route.php(259): Illuminate\\Routing\\ControllerDispatcher->dispatch()
#18 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Route.php(205): Illuminate\\Routing\\Route->runController()
#19 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(799): Illuminate\\Routing\\Route->run()
#20 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(141): Illuminate\\Routing\\Router->Illuminate\\Routing\\{closure}()
#21 /var/www/html/app/Http/Middleware/IsReadyForStep.php(54): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#22 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): App\\Http\\Middleware\\ServiceControllerMiddleware->handle()
#23 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php(50): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#24 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Routing\\Middleware\\SubstituteBindings->handle()
#25 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php(78): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#26 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\VerifyCsrfToken->handle()
#27 /var/www/html/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php(49): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#28 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\View\\Middleware\\ShareErrorsFromSession->handle()
#29 /var/www/html/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(121): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#30 /var/www/html/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(64): Illuminate\\Session\\Middleware\\StartSession->handleStatefulRequest()
#31 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Session\\Middleware\\StartSession->handle()
#32 /var/www/html/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php(37): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#33 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Cookie\\Middleware\\AddQueuedCookiesToResponse->handle()
#34 /var/www/html/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php(67): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#35 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Cookie\\Middleware\\EncryptCookies->handle()
#36 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(116): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#37 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(798): Illuminate\\Pipeline\\Pipeline->then()
#38 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(777): Illuminate\\Routing\\Router->runRouteWithinStack()
#39 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(741): Illuminate\\Routing\\Router->runRoute()
#40 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(730): Illuminate\\Routing\\Router->dispatchToRoute()
#41 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(200): Illuminate\\Routing\\Router->dispatch()
#42 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(141): Illuminate\\Foundation\\Http\\Kernel->Illuminate\\Foundation\\Http\\{closure}()
#43 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#44 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php(31): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle()
#45 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\ConvertEmptyStringsToNull->handle()
#46 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#47 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php(40): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle()
#48 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\TrimStrings->handle()
#49 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#50 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize->handle()
#51 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php(86): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#52 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Foundation\\Http\\Middleware\\PreventRequestsDuringMaintenance->handle()
#53 /var/www/html/vendor/laravel/framework/src/Illuminate/Http/Middleware/HandleCors.php(49): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#54 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Http\\Middleware\\HandleCors->handle()
#55 /var/www/html/vendor/laravel/framework/src/Illuminate/Http/Middleware/TrustProxies.php(39): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#56 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(180): Illuminate\\Http\\Middleware\\TrustProxies->handle()
#57 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(116): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#58 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(175): Illuminate\\Pipeline\\Pipeline->then()
#59 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(144): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter()
#60 /var/www/html/public/index.php(76): Illuminate\\Foundation\\Http\\Kernel->handle()
#61 {main}
"} 
192.168.10.50 - - [15/Jun/2023:18:11:12 +0000] "GET /validate/nordigen HTTP/1.1" 500 1144 "http://192.168.10.100:8098/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.5 Safari/605.1.15"
```
I am running FIDI as a container in a stack, using the NORDIGEN_ID and NORDIGEN_KEY environment variable in the stack.

I have tried to:
- use curl from my mac to authenticate from the terminal as suggested by Gocardless (ex Nordigen) https://bankaccountdata.gocardless.com/user-secrets/ : it works
- use curl from the host machine where FIDI runs (so I am sure it has the right IP address) : it works
- use curl from within the container terminal : it works
- authenticate using ob.nordigen.com/api/v2/token/new instead of bankaccountdata.gocardless.com/api/v2/token/new which is suggested in gocardless page
- rebuild the container from scratch (redeployed the stack)


But FIDI stil cannot authenticate to nordigen.
Does FIDI saves **somewhere** the old credentials??

**Reply by JC5** [reply]
So what happens when you press Reauthenticate?

**Reply by nick2k3** [reply]
the log above is the one after I click Reauthenticate

**Comment by nick2k3** [comment]
weird! I tried using a private windows and... it works!
<img width="1439" alt="Screenshot 2023-06-15 alle 23 13 51" src="https://github.com/firefly-iii/firefly-iii/assets/42848864/2971c470-6d7b-41ad-84dc-18959589bd29">

so it seems that somehow FIDI is storing my keys in the cookies?

**Reply by JC5** [reply]
Yes it is :)

**Reply by rchl** [reply]
As I said above: https://github.com/orgs/firefly-iii/discussions/7237#discussioncomment-6136605

**Reply by nick2k3** [reply]
I mistakenly interpreted as some cookies from FIDI towards nordigen API :/
is it really necessary to have the session ID and token both in the configuration and the cookie?
security concerns aside (if one has access to those cookies i have bigger problems ) it's a bit hard to find out that those cookies exist at all..

**Reply by JC5** [reply]
It's a stateless container, I'm open to ideas. I think I will add something to the data importer itself because clearly nobody RTFM.

**Reply by nick2k3** [reply]
I see, could the ID and secret be gathered from the environment variables while the token is stored in the session (cookie)?
Otherwise, maybe flush these keys in the cookies when the Reauthenticate button is pressed?

**Reply by JC5** [reply]
Alas, all of these are already in place. See also the code for [Reauthenticate](https://github.com/firefly-iii/data-importer/blob/main/app/Http/Controllers/IndexController.php#L133). As such, I'm still looking for a way to make sure this works flawlessly.

**Reply by rchl** [reply]
As I've mentioned before, the Nordigen auth cookies are not cleared on pressing Reauthenticate.

**Reply by JC5** [reply]
What happens when you browse to /flush instead?

**Reply by rchl** [reply]
It redirects back to importer and with no perceptible change (still authenticated to firefly and nordigen)

**Reply by JC5** [reply]
Mmmm. Keep in mind that when the index loads (without cookies) it _will_ try to reauthenticate, it just guarantees (or tries to) it uses the latest versions of all the keys and secrets. Your logs will reveal what's happening.

**Comment by nick2k3** [comment]
I just noticed that, before manually removing my cookies, FIDI was saving also client id and secret along with the tooken.
After flushing and doing one importation, I only see the token.
Perhaps older version of FIDI stored them and newer don't?
<img width="254" alt="Screenshot 2023-06-16 alle 11 34 57" src="https://github.com/firefly-iii/firefly-iii/assets/42848864/bb2e51a5-0c90-400d-a29e-af63e18e0aa7">

**Reply by JC5** [reply]
Is your data importer sharing the subdomain with firefly iii?

**Reply by nick2k3** [reply]
I guess so, they don't have a proper domain name, I just access them by IP:port
